src/core/tsi/s2a/handshaker/s2a_handshaker_client_caller.cc

```diff
@@ -30,16 +32,208 @@ namespace experimental {
  *  function of the |S2AHandshakerClient| class. This method enables the S2A
  *  handshaker client to make a gRPC call to the S2A service. **/
 
+const size_t kHandshakerClientOpNum = 4;
+
+/** The implementation of |make_grpc_call| is nearly identical to that of its
+ *  ALTS counterpart, the |make_grpc_call| method in alts_handshaker_client.cc.
+ */
 tsi_result S2AHandshakerClient::MakeGrpcCall(bool is_start) {
-  // TODO(mattstev): implement.
-  return TSI_UNIMPLEMENTED;
+  grpc_op ops[kHandshakerClientOpNum];
+  memset(ops, 0, sizeof(ops));
+  grpc_op* op = ops;
+  if (is_start) {
+    op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
+    op->data.recv_status_on_client.trailing_metadata = nullptr;
+    op->data.recv_status_on_client.status = &handshake_status_code_;
+    op->data.recv_status_on_client.status_details = &handshake_status_details_;
+    op->flags = 0;
+    op->reserved = nullptr;
+    op++;
+    GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+    gpr_ref(refs_);
+    GPR_ASSERT(grpc_caller_ != nullptr);
+    grpc_call_error call_error = grpc_caller_(
+        call_, ops, static_cast<size_t>(op - ops), &on_status_received_);
+    GPR_ASSERT(call_error == GRPC_CALL_OK);
+    memset(ops, 0, sizeof(ops));
+    op = ops;
+    op->op = GRPC_OP_SEND_INITIAL_METADATA;
+    op->data.send_initial_metadata.count = 0;
+    op++;
+    GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+    op->op = GRPC_OP_RECV_INITIAL_METADATA;
+    op->data.recv_initial_metadata.recv_initial_metadata =
+        &recv_initial_metadata_;
+    op++;
+    GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+  }
+  op->op = GRPC_OP_SEND_MESSAGE;
+  op->data.send_message.send_message = send_buffer_;
+  op++;
+  GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+  op->op = GRPC_OP_RECV_MESSAGE;
+  op->data.recv_message.recv_message = &recv_buffer_;
+  op++;
+  GPR_ASSERT(op - ops <= kHandshakerClientOpNum);
+  GPR_ASSERT(grpc_caller_ != nullptr);
+  if (grpc_caller_(call_, ops, static_cast<size_t>(op - ops),
+                   &on_handshaker_service_resp_recv_) != GRPC_CALL_OK) {
+    gpr_log(GPR_ERROR, "Start batch operation failed");
+    return TSI_INTERNAL_ERROR;
+  }
+  return TSI_OK;
 }
 
+/** The implementation of this method is nearly identical to its ALTS
```

Comment:
Not sure if we want to keep this comment especially if we do not want to place the code in gRPC repo.

Reply:
Removed in next commit.